File: src/main/java/com/glitch/knowledge/model/StationToLine.java
package com.glitch.knowledge.model;

import com.glitch.knowledge.model.constant.StationDirection;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

/**
 * Map of {@link Station} to {@link Line}.
 * <br>
 * There should be one entry for every line a station might service. At least one entry per station.
 * <br>
 * Also includes specific information of the station for any associated line.
 */
@Entity(name = "station_line")
public class StationToLine {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @ManyToOne
    @JoinColumn(name = "station_id", referencedColumnName = "id")
    private Station station;

    @ManyToOne
    @JoinColumn(name = "line_id", referencedColumnName = "id")
    private Line line;

    @Column(nullable = false)
    private int position;

    @Column(nullable = true)
    private Integer branch_id;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private StationDirection direction;

    /**
     * @return Mapped {@link Line} object
     */
    public Line getLine() {
        return this.line;
    }

    /**
     * @return Arbitrarily chosen station weight on the line. Starting station is 0 and increases to the end.
     */
    public int getPosition() {
        return this.position;
    }

    /**
     * @return If the station on this line is on a branch of the line
     */
    public boolean isBranch() {
        return this.branch_id != null;
    }

    /**
     * @return {@link Station} ID of the branch (terminus of the branch). Is null if the station is not on a branch.
     */
    public Integer getBranchId() {
        return this.branch_id;
    }

    /**
     * @return {@link StationDirection} of the station on the associated line
     */
    public StationDirection getDirection() {
        return this.direction;
    }
}

File: src/main/kotlin/com/glitch/knowledge/model/StationToLine.kt
package com.glitch.knowledge.model

import com.glitch.knowledge.model.constant.StationDirection
import org.jetbrains.exposed.v1.core.dao.id.EntityID
import org.jetbrains.exposed.v1.core.dao.id.IntIdTable
import org.jetbrains.exposed.v1.dao.ImmutableEntityClass
import org.jetbrains.exposed.v1.dao.IntEntity


internal object StationToLineTable : IntIdTable("station_line") {
    val stationId = reference("station_id", StationTable.id)
    val lineId = reference("line_id", LineTable.id)
    val position = integer("position")
    val branchId = integer("branch_id").nullable()
    val direction = enumerationByName("direction", 16, StationDirection::class)
}


internal class StationToLineDAO(id: EntityID<Int>) : IntEntity(id) {
    companion object : ImmutableEntityClass<Int, StationToLineDAO>(StationToLineTable)

    val stationId by StationToLineTable.stationId
    val lineId by StationToLineTable.lineId

    val station by StationDAO referencedOn StationToLineTable.stationId
    val line by LineDAO referencedOn StationToLineTable.lineId

    val position: Int by StationToLineTable.position
    val branchId: Int? by StationToLineTable.branchId
    val direction: StationDirection by StationToLineTable.direction
}


/**
 * Map of [Station] to [Line].
 *
 * There should be one entry for every line a station might service. At least
 * one entry per station.
 *
 * Also includes specific information of the station for any associated line.
 * @property stationId ID of the [Station] being mapped.
 * @property lineId ID of the [Line] being mapped.
 * @property position Arbitrarily chosen [Station] weight on the [Line]. Starting station has position 0 and increases
 * to the end.
 * @property branchId [Station.id] of the [Terminus] of the [Line], if the mapped station is on a branch of that line.
 * @property direction [StationDirection] of the mapped [Station] on the associated [Line].
 */
class StationToLine(
    val stationId: Int,
    val lineId: Int,
    val position: Int,
    val branchId: Int?,
    val direction: StationDirection
) {
    companion object {
        internal fun fromDAO(dao: StationToLineDAO): StationToLine {
            return StationToLine(
                dao.stationId.value, 
                dao.lineId.value, 
                dao.position, 
                dao.branchId, 
                dao.direction
            )
        }
    }

    /**
     * If the station on this line is on a branch of the line
     */
    fun isBranch(): Boolean {
        return this.branchId != null
    }
}
